src/macaron/slsa_analyzer/ci_service/github_actions.py

```diff
@@ -410,6 +412,116 @@ def has_latest_run_passed(
 
         return ""
 
+    def workflow_run_in_date_time_range(
+        self,
+        repo_full_name: str,
+        workflow: str,
+        date_time: datetime,
+        step_name: str,
+        time_range: int = 0,
+    ) -> set[str]:
+        """Check if the repository has a workflow run started before the date_time timestamp within the time_range.
+
+        - This method queries the list of workflow runs using the GitHub API for the provided repository full name.
+        - It will filter out the runs that are not triggered by the given workflow.
+        - It will only accept the runs that from `date_time - time_range` to `date_time`.
+        - If a `step_name` is provided, checks that it has started before the `date_time` and has succeeded.
+
+        Parameters
+        ----------
+        repo_full_name : str
+            The target repo's full name.
+        workflow : str
+            The workflow URL.
+        date_time: datetime
+            The datetime object to query.
+        step_name: str
+            The step in the GitHub Action workflow that needs to be checked.
+        time_range: int
+            The date-time range in seconds. The default value is 0.
+            For example a 30 seconds range for 2022-11-05T20:30 is 2022-11-05T20:15..2022-11-05T20:45.
+
+        Returns
+        -------
+        set[str]
+            The set of URLs found for the workflow within the time range.
+        """
+        logger.debug(
+            "Getting the latest workflow run of %s at %s within time range %s",
+            workflow,
+            str(date_time),
+            str(time_range),
+        )
+
+        html_urls: set[str] = set()
+        try:
+            datetime_from = date_time - timedelta(seconds=time_range)
+        except (OverflowError, OSError, TypeError) as error:
+            logger.debug(error)
+            return html_urls
+
+        # Perform the search.
+        logger.debug("Search for the workflow runs within the range.")
+        try:
+            run_data = self.api_client.get_workflow_run_for_date_time_range(
+                repo_full_name, f"{datetime_from.isoformat()}..{date_time.isoformat()}"
+            )
+        except ValueError as error:
+            logger.debug(error)
+            return html_urls
+
+        if not run_data:
+            logger.debug("Unable to find any run data for the workflow %s", workflow)
+            return html_urls
+
+        logger.debug("Checking workflow run of %s.", workflow)
+        try:
+            # iterate through the responses in reversed order to add the run closest to the date_time timestamp first.
+            for item in reversed(run_data["workflow_runs"]):
```

Comment:
From my understanding, the list of workflow runs when returned by the REST API will be in the order of **decreasing** created date. This will mean that the latest workflow run will be at the first position in `run_data["workflow_runs"]`, etc. 
However, from the comment here
```
# iterate through the responses in reversed order to add the run closest to the date_time timestamp first.
```
If we are checking them in the reversed order, the workflow runs closest to `date_time-time_range` will be encountered first. I think there is an inconsistent between the comment and the implementation.

Reply:
>If we are checking them in the reversed order, the workflow runs closest to date_time-time_range will be encountered first. 

This workflow has happened earlier and appears as the last element in the `run_data["workflow_runs"]`. We want to prioritize it and show it as the first candidate in the justification, so we need to iterate in a reversed order.

Reply:
If we are prioritizing the earliest workflow run in the range of `date_time - time_range` to `date_time` (given that `date_time - time_range` is earlier than `date_time`) then should the comment here be:
```
# iterate through the responses in reversed order to add the run closest to the ``date_time - time_range`` timestamp first.
```

Reply:
See commit [a656eb7](https://github.com/oracle/macaron/pull/471/commits/a656eb719141173d47f473ea1fc1607cab610525)